docs/architecture/adr-001-node-key-refactoring.md

```diff
@@ -0,0 +1,104 @@
+# ADR ADR-001: Node Key Refactoring
+
+## Changelog
+
+- 2022-10-31: First draft
+
+## Status
+
+Proposed
+
+## Context
+
+The original key format of IAVL nodes is a hash of the node. It does not take advantage of data locality on disk. Nodes are stored in random locations on disk due to the random hash value, so it needs to scan the disk to find the corresponding node which can be very inefficient.
```

Comment:
Does the "data locality" make any significant difference in SSD disks?

Reply:
Moreover, the data layout will depend on the underlying DB structure (goleveldb, rocksdb...) -- and most likely it will be different then we think. 
Do we have a benchmark proving the statement?

Reply:
https://github.com/cosmos/iavl/pull/597#issuecomment-1302490107
maybe it doesn't reflect fully the above decisions, it only refactored the nodeKey as a global nonce.

Reply:
Data locality makes a difference in any LSM design.

The reason is that an LSM key-value store performs compactions under the hood to remove duplicated entries

Reply:
Robert, data locality is the central problem for all computational work, including CPU caches, page tables and RAM.

Please read up on caches and the memory hierarchy.

Reply:
I think "data locality" maybe a bit misleading here, because even with the new key format, we don't really have data locality for most of cases, because a node's children are likely belong to different versions, so they are still located distantly.
For me, the biggest advantage of new format is, the key is monotonically increasing, so it works better with kv-dbs where most of them store data in some kind of ordered trees, and our inserted keys are already ordered, and never modify(ignore pruning for now). We can even simply [dump the nodes into plain files](https://github.com/cosmos/iavl/pull/608#issuecomment-1301547758), they can still be queryable with some minimal metadata.